server/etcdserver/dump.go

```diff
@@ -0,0 +1,653 @@
+// Copyright 2021 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdserver
+
+import (
+	"bufio"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type RPCRequestType int
+type RPCResponseType int
+
+const (
+	RPCReqTypeUnknown = RPCRequestType(0)
+	RPCReqTypePut     = RPCRequestType(1)
+	RPCReqTypeRange   = RPCRequestType(2)
+	RPCReqTypeTxn     = RPCRequestType(3)
+	RPCReqTypeDelete  = RPCRequestType(4)
+	RPCReqTypeCompact = RPCRequestType(5)
+)
+
+const (
+	RPCRespTypeUnknown = RPCResponseType(0)
+	RPCRespTypePut     = RPCResponseType(1)
+	RPCRespTypeRange   = RPCResponseType(2)
+	RPCRespTypeTxn     = RPCResponseType(3)
+	RPCRespTypeDelete  = RPCResponseType(4)
+	RPCRespTypeCompact = RPCResponseType(5)
+)
+
+const (
+	GRPCEntriesBufMax   = 10000
+	HTTPPrefixGRPCDebug = "/debug"
+)
+
+// GRPCDumpEntry is the data inside entriesCh
+type GRPCDumpEntry struct {
+	Start        int64
+	End          int64
+	Method       string
+	Metadata     metadata.MD
+	RequestType  RPCRequestType
+	Request      interface{}
+	ResponseType RPCResponseType
+	Response     interface{}
+}
+
+// GRPCDumpData is the actual type of data saved in dump file
+type GRPCDumpData struct {
+	Start        int64
+	End          int64
+	Method       string
+	Metadata     metadata.MD
+	RequestType  RPCRequestType
+	Request      []byte
+	ResponseType RPCResponseType
+	Response     []byte
+}
+
+var (
+	// dump data channel
```

Comment:
Can we define the these below within struct, to limit access and make it more testable? /cc @lilic